sys/net/link_layer/ieee802154/security.c

```diff
@@ -0,0 +1,521 @@
+       ACKs are not encrypted. */
+    assert((*((uint8_t *)header)) & IEEE802154_FCF_TYPE_DATA);
+
+    if (ctx->security_level == IEEE802154_SCF_SECLEVEL_NONE) {
+        *mic_size = 0;
+        return IEEE802154_SEC_OK;
+    }
+    if (ctx->frame_counter == 0xFFFFFFFF) {
+        /* Letting the frame counter overflow is explicitly prohibited by the specification.
+           (see 9.4.2) */
+        return -IEEE802154_SEC_FRAME_COUNTER_OVERFLOW;
+    }
+
+    /* write the auxiliary header */
+    ieee802154_aux_sec_t *aux = (ieee802154_aux_sec_t *)(header + *header_size);
+    uint8_t aux_size = _get_aux_hdr_size(ctx->security_level, ctx->key_id_mode);
+    _set_aux_hdr(ctx, aux);
+
+    /* attempt to find the encrypton key */
+    const uint8_t *key;
+    if (!(key = _get_encryption_key(ctx, header, *header_size, aux))) {
+        return -IEEE802154_SEC_NO_KEY;
+    }
+    _set_key(ctx, key);
+
+    *mic_size = _mac_size(ctx->security_level);
+    const uint8_t *a = header;
+    uint8_t *m = payload;
+    uint16_t a_len = *header_size + aux_size;
+    uint16_t m_len = payload_size;
+    ieee802154_ccm_block_t ccm; /* Ai or Bi */
+
+    /* compute MIC */
+    if (_req_mac(ctx->security_level)) {
+        _init_cbc_B0(&ccm, ctx->frame_counter, ctx->security_level, m_len, *mic_size, src_address);
+        _comp_mic(ctx, mic, &ccm, a, a_len, m, m_len);
+
+        /* encrypt MIC */
+        _init_ctr_A0(&ccm, ctx->frame_counter, ctx->security_level, src_address);
+        _ctr_mic(ctx, &ccm, mic, *mic_size);
+    }
+    /* encrypt payload */
+    if (_req_encryption(ctx->security_level)) {
+        _init_ctr_A0(&ccm, ctx->frame_counter, ctx->security_level, src_address);
+        _ctr(ctx, &ccm, m, m_len);
+    }
+    *header_size += aux_size;
+    ctx->frame_counter++;
+    return IEEE802154_SEC_OK;
+}
+
+int ieee802154_sec_decrypt_frame(ieee802154_sec_context_t *ctx,
+                                 uint16_t frame_size,
+                                 uint8_t *header, uint8_t *header_size,
+                                 uint8_t **payload, uint16_t *payload_size,
+                                 uint8_t **mic, uint8_t *mic_size,
+                                 const uint8_t *src_address)
+{
+    /* For non data frames (MAC commands, beacons) a and a_len would be larger.
+       ACKs are not encrypted. */
+    assert(*header & IEEE802154_FCF_TYPE_DATA);
+
+    /* read the fields of the auxiliary header */
+    ieee802154_aux_sec_t *aux = (ieee802154_aux_sec_t *)(header + *header_size);
+    uint8_t security_level = _get_sec_level(aux->scf);
+    uint8_t key_mode = _get_key_id_mode(aux->scf);
+    uint8_t aux_size = _get_aux_hdr_size(security_level, key_mode);
+    uint8_t mac_size = _mac_size(security_level);
+    /* remember that the frame counter was stored in little endian */
+    uint32_t frame_counter = byteorder_ntohl((be_uint32_t)
+                                byteorder_ltobl((le_uint32_t)aux->fc));
```

Comment:
gcc 10 does not like this

```
/home/benpicco/dev/RIOT-BHT/sys/net/link_layer/ieee802154/security.c: In function ‘ieee802154_sec_decrypt_frame’:
/home/benpicco/dev/RIOT-BHT/sys/net/link_layer/ieee802154/security.c:426:49: error: ISO C forbids casts to union type [-Werror=pedantic]
  426 |                                 byteorder_ltobl((le_uint32_t)aux->fc));
      |                                                 ^
/home/benpicco/dev/RIOT-BHT/sys/net/link_layer/ieee802154/security.c:425:46: error: ISO C forbids casting nonscalar to the same type [-Werror=pedantic]
  425 |     uint32_t frame_counter = byteorder_ntohl((be_uint32_t)
      |                                              ^
cc1: all warnings being treated as errors
```
Given that this code will likely never run on a big endian machine, we might as well do
```suggestion
    uint32_t frame_counter = aux->fc;
```

but a 'proper' fix would be 

```suggestion
    uint32_t frame_counter = byteorder_ntohl(byteorder_ltobl((le_uint32_t){aux->fc}));
````

Reply:
I find your lack of future-proofing disturbing ;-P

Reply:
I'm merely calling the future of Big Endian machines into question ;)
But I remember someone on IRC mentioning that they were working on an AVR32 port for RIOT. Granted, that also falls into the realm of retro-computing, but it would be a Big Endian architecture after all.

Anyway, there is a proper way to get rid of the compiler warning and the compiler should be clever enough to turn this into a NOP on little endian.